tpdatasrc/tpgamefiles/scr/spell_utils.py

```diff
@@ -0,0 +1,275 @@
+from templeplus.pymod import PythonModifier
+from toee import *
+import tpdp
+from utilities import *
+
+# Get the spell's name from its spell_id
+def spellName(spellId):
+    spellEnum = tpdp.SpellPacket(spellId).spell_enum
+    return game.get_spell_mesline(spellEnum)
+
+def spellKeyName(spellId):
+    return spellName(spellId).upper().replace(" ", "_")
+
+def spellKey(spellId):
+    return tpdp.hash(spellKeyName(spellId))
+
+def spellTime(duration):
+    if duration == 1:
+        return "1 round"
+    elif duration < 100: # 10 minutes
+        return "{} rounds".format(duration)
+    elif duration < 1200: # 2 hours
+        return "{} minutes".format(duration/10)
+    elif duration < 28800: # 2 days
+        return "{} hours".format(duration/600)
+    elif duration < 864000: # 2 months
+        return "{} days".format(duration/14400)
+    else:
+        return "{} months".format(duration/432000)
+
+### Standard Hooks invoked with AddHook ###
+
+#[pytonModifier].AddHook(ET_OnGetTooltip, EK_NONE, spell_utils.spellTooltip, ())
+def spellTooltip(attachee, args, evt_obj):
+    name = spellName(args.get_arg(0))
+    duration = spellTime(args.get_arg(1))
+    evt_obj.append("{} ({})".format(name, duration))
+    return 0
+
+#[pytonModifier].AddHook(ET_OnGetEffectTooltip, EK_NONE, spell_utils.spellEffectTooltip, ())
+def spellEffectTooltip(attachee, args, evt_obj):
+    key = spellKey(args.get_arg(0))
+    duration = spellTime(args.get_arg(1))
+    evt_obj.append(key, -2, " ({})".format(duration))
+    return 0
+
+#[pytonModifier].AddHook(ET_OnD20Query, EK_Q_Critter_Has_Spell_Active, spell_utils.queryActiveSpell, ())
+def queryActiveSpell(attachee, args, evt_obj):
+    spellPacket = tpdp.SpellPacket(args.get_arg(0))
+    if evt_obj.data1 == spellPacket.spell_enum:
+        evt_obj.return_val = 1
+    return 0
+
+#[pytonModifier].AddHook(ET_OnD20Signal, EK_S_Killed, spell_utils.spellKilled, ())
+def spellKilled(attachee, args, evt_obj):
+    args.remove_spell()
+    args.remove_spell_mod()
+    return 0
+
+#[pytonModifier].AddHook(ET_OnD20Signal, EK_S_Spell_End, spell_utils.spellEnd, ())
+def spellEnd(attachee, args, evt_obj):
+    print "{} SpellEnd".format(spellName(args.get_arg(0)))
```

Comment:
Shouldn't spellEnd actually do something? The common spell ender does various things.
@Sagenlicht

Reply:
This is just an additional hook that prints to the console (the end of the spell). I just adopted it from spells like Moment of Prescience. I rarely do anything on Spell End in my spells, e.g. spell particles that are permanent are not cleared by my spells but the game ends them. I always assumed (and still do!) that the game ends all particles registered within a spell_id when the spell ends.
Long story short, Close to all my spells have this exact hook with just a print to the console. If this is not required, I think we can just drop the hook and only use it, when it is actually required to do something in spellEnd.
An example where I actually do something in spellEnd would be Iron Guts that adds a condition once the spell expires.

Reply:
Right, I mixed this up with remove_spell which has a default handler, which together with the generic AoE handler is probably good enough for 99% of cases.

And yeah, if all it does is print to console, it can be safely removed, since you'll get a log message from remove_spell anyway.

Reply:
Ok then let's remove it :)